RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h

```diff
@@ -0,0 +1,220 @@
+
+      int nClusInIter = std::min(MaxHitsInIter, endClus - startClus);
+      int lastClus = startClus + nClusInIter;
+      assert(nClusInIter <= nclus);
+      assert(nClusInIter > 0);
+      assert(lastClus <= nclus);
+
+      assert(nclus > MaxHitsInIter || (0 == startClus && nClusInIter == nclus && lastClus == nclus));
+
+      // init
+      for (int ic = threadIdx.x; ic < nClusInIter; ic += blockDim.x) {
+        clusParams.minRow[ic] = std::numeric_limits<uint32_t>::max();
+        clusParams.maxRow[ic] = 0;
+        clusParams.minCol[ic] = std::numeric_limits<uint32_t>::max();
+        clusParams.maxCol[ic] = 0;
+        clusParams.charge[ic] = 0;
+        clusParams.Q_f_X[ic] = 0;
+        clusParams.Q_l_X[ic] = 0;
+        clusParams.Q_f_Y[ic] = 0;
+        clusParams.Q_l_Y[ic] = 0;
+      }
+
+      first += threadIdx.x;
+
+      __syncthreads();
+
+      // one thead per "digi"
+
+      for (int i = first; i < numElements; i += blockDim.x) {
+        auto id = digis.moduleInd(i);
+        if (id == InvId)
+          continue;  // not valid
+        if (id != me)
+          break;  // end of module
+        auto cl = digis.clus(i);
+        if (cl < startClus || cl >= lastClus)
+          continue;
+        auto x = digis.xx(i);
+        auto y = digis.yy(i);
+        cl -= startClus;
+        assert(cl >= 0);
+        assert(cl < MaxHitsInIter);
+        atomicMin(&clusParams.minRow[cl], x);
+        atomicMax(&clusParams.maxRow[cl], x);
+        atomicMin(&clusParams.minCol[cl], y);
+        atomicMax(&clusParams.maxCol[cl], y);
+      }
+
+      __syncthreads();
+
+      auto pixmx = cpeParams->detParams(me).pixmx;
+      for (int i = first; i < numElements; i += blockDim.x) {
+        auto id = digis.moduleInd(i);
+        if (id == InvId)
+          continue;  // not valid
+        if (id != me)
+          break;  // end of module
+        auto cl = digis.clus(i);
+        if (cl < startClus || cl >= lastClus)
+          continue;
+        cl -= startClus;
+        assert(cl >= 0);
+        assert(cl < MaxHitsInIter);
+        auto x = digis.xx(i);
+        auto y = digis.yy(i);
+        auto ch = std::min(digis.adc(i), pixmx);
+        atomicAdd(&clusParams.charge[cl], ch);
+        if (clusParams.minRow[cl] == x)
+          atomicAdd(&clusParams.Q_f_X[cl], ch);
+        if (clusParams.maxRow[cl] == x)
+          atomicAdd(&clusParams.Q_l_X[cl], ch);
+        if (clusParams.minCol[cl] == y)
+          atomicAdd(&clusParams.Q_f_Y[cl], ch);
+        if (clusParams.maxCol[cl] == y)
+          atomicAdd(&clusParams.Q_l_Y[cl], ch);
+      }
+
+      __syncthreads();
+
+      // next one cluster per thread...
+
+      first = clusters.clusModuleStart(me) + startClus;
+
+      for (int ic = threadIdx.x; ic < nClusInIter; ic += blockDim.x) {
+        auto h = first + ic;  // output index in global memory
+
+        // this cannot happen anymore
+        if (h >= TrackingRecHit2DSOAView::maxHits())
+          break;  // overflow...
```

Comment:
if it cannot happen, replace with an assert?

Reply:
Probably makes sense.

A general comment: `assert` in device code are extremely slow, so they are disabled by default (unlike `assert` in host code that are enabled by default in CMSSW).

Reply:
Changes to an `assert` in https://github.com/cms-patatrack/cmssw/pull/593/commits/a12321f7ab1da5dd1d3b34043c660db4809aafb1 .